packages/core/src/hooks/useReadContract.ts

```diff
@@ -0,0 +1,109 @@
+import { ArgsOrCalldata, BlockNumber } from "starknet";
+
+import { UseQueryProps, UseQueryResult } from "~/query";
+
+import {
+  Abi,
+  ExtractAbiFunctionNames,
+  FunctionArgs,
+  FunctionRet,
+} from "abi-wan-kanabi/dist/kanabi";
+
+import { CallQueryKey, UseCallProps, useCall } from "./useCall";
+
+type ReadContractArgs = {
+  /** The contract's function name. */
+  functionName: string;
+  /** Read arguments. */
+  args?: ArgsOrCalldata;
+  /** Block identifier used when performing call. */
+  blockIdentifier?: BlockNumber;
+};
+
+type Result<
+  TAbi extends Abi,
+  TFunctionName extends ExtractAbiFunctionNames<TAbi>
+> = FunctionRet<TAbi, TFunctionName>;
+
+/** Options for `useReadContract`. */
+export type UseReadContractProps<
+  TAbi extends Abi,
+  TFunctionName extends ExtractAbiFunctionNames<TAbi>
+> = Omit<ReadContractArgs, "functionName" | "args"> &
+  UseQueryProps<
+    Result<TAbi, TFunctionName>,
+    Error,
+    Result<TAbi, TFunctionName>,
+    ReturnType<CallQueryKey>
+  > & {
+    /** The target contract's ABI.
+     *
+     * @remarks
+     *
+     * You must pass ABI as const
+     *
+     * @example
+     * abi: [
+     *   {
+     *     type: "function",
+     *     name: "fn_simple_array",
+     *     inputs: [
+     *       {
+     *         name: "arg",
+     *         type: "core::array::Array::<core::integer::u8>",
+     *       },
+     *     ],
+     *     outputs: [],
+     *     state_mutability: "view",
+     *   }
+     *  ] as const
+     *
+     */
+    abi?: TAbi;
+    /** The target contract's address. */
+    address?: string;
+    /** Refresh data at every block. */
+    watch?: boolean;
+    /** The contract's function name. */
+    functionName: TFunctionName;
+    /** Read arguments. */
+    args?: FunctionArgs<TAbi, TFunctionName>;
+  };
```

Comment:
```suggestion
> = UseQueryProps<
    Result<TAbi, TFunctionName>,
    Error,
    Result<TAbi, TFunctionName>,
    ReturnType<CallQueryKey>
  > & {
    /** Block identifier used when performing call. */
  blockIdentifier?: BlockNumber;
    /** The target contract's ABI.
     *
     * @remarks
     *
     * You must pass ABI as const
     *
     * @example
     * abi: [
     *   {
     *     type: "function",
     *     name: "fn_simple_array",
     *     inputs: [
     *       {
     *         name: "arg",
     *         type: "core::array::Array::<core::integer::u8>",
     *       },
     *     ],
     *     outputs: [],
     *     state_mutability: "view",
     *   }
     *  ] as const
     *
     */
    abi?: TAbi;
    /** The target contract's address. */
    address?: string;
    /** Refresh data at every block. */
    watch?: boolean;
    /** The contract's function name. */
    functionName: TFunctionName;
    /** Read arguments. */
    args?: [...ExtractArgs<TAbi, ExtractAbiFunction<TAbi, TFunctionName>>],
  };
```

Let's add the `blockIdentifier` in the main args directly. Also I think I've found a way to always enforce an array of parameters. Need to import some auxiliary functions from abi wan.